src/lib/storage/constraints/base_constraint_checker.hpp

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include <optional>
+#include <string>
+#include <tuple>
+
+#include "storage/chunk.hpp"
+#include "storage/table.hpp"
+
+namespace opossum {
+
+class BaseConstraintChecker {
+ public:
+  BaseConstraintChecker(const Table& table, const TableConstraintDefinition& constraint)
+      : _table(table), _constraint(constraint) {}
+  virtual ~BaseConstraintChecker() = default;
+
+  virtual std::tuple<bool, ChunkID> isValid(const CommitID snapshot_commit_id, const TransactionID our_tid) = 0;
+  virtual std::tuple<bool, ChunkID> isValidForInsertedValues(std::shared_ptr<const Table> table_to_insert,
+                                                             const CommitID snapshot_commit_id,
+                                                             const TransactionID our_tid, const ChunkID since) = 0;
+
+ protected:
+  const Table& _table;
+  const TableConstraintDefinition& _constraint;
+};
+
+template <typename Row>
+class BaseBaseConstraintChecker : public BaseConstraintChecker {
+ public:
+  BaseBaseConstraintChecker(const Table& table, const TableConstraintDefinition& constraint)
+      : BaseConstraintChecker(table, constraint) {}
+
+  virtual std::shared_ptr<std::vector<Row>> getInsertedRows(std::shared_ptr<const Table> table) const = 0;
+
+  /**
+   * Prepare for returning values when getRow is called.
+   * Return false if the segment doesn't need to be checked for duplicate values.
+   * TODO(pk_group): naming
+   */
+  virtual bool prepareReadChunk(std::shared_ptr<const Chunk> chunk) {
```

Comment:
Yeah the naming and return type meaning... We could also split it into two methods but the segment/s are taken from the chunk nicely in the `prepareReadChunk` method and so the dictionary segment check can be done there immediately without implicitly assuming that this method was called before.